### PR TITLE
Support `recvmsg()` and `sendmsg()`

### PR DESCRIPTION
**Pull request by lum1n0us**
Add an example, *send_recv.c*, to test with new socket APIs. Before testing, please uncomment lines in CMakeLists.txt

### CODE REVIEW

**Comment by lum1n0us** [comment]
@wenyongh  what kind of copyright declaration should I add into [the example](samples/socket-api/wasm-src/send_recv.c) ?

**Comment by lum1n0us** [comment]
@wenyongh here is the question,  the new API `socketpair()`
> On Linux, the only supported domains for this call are AF_UNIX (or synonymously, AF_LOCAL) and AF_TIPC (since Linux 4.12)

And WAMR socket supporting only covers `AF_INET`, which means either we expand the socket address family or we reject the requirement. 

BTW, the example *send_recv.c* will be failed in linux:
```
$ clang-13 -o send_recv ../wasm-src/send_recv.c
$ ./send_recv
socketpair err: Operation not supported
```

**Comment by wenyongh** [comment]
> @wenyongh here is the question, the new API `socketpair()`
> 
> > On Linux, the only supported domains for this call are AF_UNIX (or synonymously, AF_LOCAL) and AF_TIPC (since Linux 4.12)
> 
> And WAMR socket supporting only covers `AF_INET`, which means either we expand the socket address family or we reject the requirement.
> 
> BTW, the example _send_recv.c_ will be failed in linux:
> 
> ```
> $ clang-13 -o send_recv ../wasm-src/send_recv.c
> $ ./send_recv
> socketpair err: Operation not supported
> ```

Please change AF_INET to AF_LOCAL.

**Comment by lum1n0us** [comment]
In that case (`socketpair()` and `AF_UNIX`), we are looking at questions and they are all connected with rights control.
- `socketpair()` creates two ***unnamed*** sockets. It required a new way to announce "unnamed socket are allowed" since neither *--addr-pool* nor *--dir* covers it now
- because `socketpair()` brings `AF_UNIX` into WAMR socket APIs supporting range, are we going to expand `AF_UNIX` to `socket()` also?  It should be OK to use '--dir' to cover "pathname sockets". (because pathname includes files and dirs, it's better to rename `--dir` to `--path`). Just need to figure out two options to cover *unnamed sockets* (created by `socketpair()`) and *abstract sockets*.

Reference:
[AF_UNIX](https://man7.org/linux/man-pages/man7/unix.7.html)
[socketpair)(https://man7.org/linux/man-pages/man2/socketpair.2.html)

**Comment by lum1n0us** [comment]
For the record, we plan to bypass the requirement of `socketpair` and only support `recvmsg()` and `sendmsg()`.

File: samples/socket-api/wasm-src/send_recv.c [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+    if (pthread_cond_init(&cond, NULL)) {
+        perror("Initialize condition failed");
+        ret = EXIT_FAILURE;
+        goto DESTROY_MUTEX;
+    }
+
+    if (pthread_create(&cs[0], NULL, run_as_server, NULL)) {
+        perror("Create a server thread failed");
+        goto DESTROY_COND;
```
Comment (wenyongh) [inline samples/socket-api/wasm-src/send_recv.c:189]: Should set `ret = EXIT_FAILURE;` before return?

File: samples/socket-api/wasm-src/send_recv.c [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    if (pthread_create(&cs[1], NULL, run_as_client, NULL)) {
+        perror("Create a client thread failed");
+        goto DESTROY_COND;
```
Comment (wenyongh) [inline samples/socket-api/wasm-src/send_recv.c:194]: Same as above

**Pull request merged**